src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -320,6 +320,61 @@ public List<JobModel> enumerateJobs(
         .collect(Collectors.toList());
   }
 
+  private List<FlightState> getAccessibleFlights(
+      int offset, int limit, FlightFilter filter, AuthenticatedUserRequest userReq)
+      throws InterruptedException {
+    int start = 0;
+    int end = offset + limit;
+    HashMap<String, List<String>> roleMap = new HashMap<>();
+    ArrayList<FlightState> flightStateList = new ArrayList<>();
+    while (flightStateList.size() < offset + limit) {
+      List<FlightState> filterResults = stairway.getFlights(start, end, filter);
+      if (filterResults.size() == 0) {
+        break;
+      }
+      filterResults.forEach(
+          flightState -> {
+            if (userLaunchedFlight(flightState, userReq)) {
+              flightStateList.add(flightState);
+            } else {
+              FlightMap inputParameters = flightState.getInputParameters();
+              IamResourceType resourceType =
+                  inputParameters.get(
+                      JobMapKeys.IAM_RESOURCE_TYPE.getKeyName(), IamResourceType.class);
+              String resourceId =
+                  inputParameters.get(JobMapKeys.IAM_RESOURCE_ID.getKeyName(), String.class);
+              IamAction action =
+                  inputParameters.get(JobMapKeys.IAM_ACTION.getKeyName(), IamAction.class);
+              if (resourceType != null & resourceId != null && action != null) {
+                String key = String.format("%s-%s", resourceType, resourceId);
+                List<String> userRoles = roleMap.get(key);
+                if (userRoles == null) {
+                  userRoles = samService.listActions(userReq, resourceType, resourceId);
```

Comment:
I do think this will get chatty...do you think it would make sense to have a simple cache for this?  Maybe as a followon.  Just thinking about when we eventually hook the UI up, this might slam Sam

Reply:
Good point! I think it would be interesting to do that after some scale testing to see the before/after